Wrapped LLM as a garak generator

Nakul-Rajpal opened the pull request:
Wrapped the Python Library LLM (for  OpenAI, Anthropic’s Claude, Google’s Gemini etc) as a garak generator.

Fixes Issue #463

Nakul-Rajpal commented:
@leondz Please check

Nakul-Rajpal commented:
@leondz Should be fixed; I forgot to add the library dependency to the toml.

Nakul-Rajpal commented:
@leondz It should be good now? The tests were failing due to me not adding the llm library to the requirements so it ran without the module.

Nakul-Rajpal commented:
@leondz Should be ready to go now

leondz commented:
Yeah, it's in the review queue, thank you

jmartin-tech reviewed (changes requested):
This looks like a great start. I noted a few edge case and a mismatch in how a `system_prompt` is handled.

Please take a look, happy to offer further detail or answer question about how things flow.

jmartin-tech commented on garak/generators/llm.py:
```diff
@@ -0,0 +1,120 @@
+
+        super().__init__(self.name, config_root=config_root)
+
+        self._clear_client()
```
It may be best to call `super().__init__()` before calling `_load_client()` to ensure the end object state is not impacted. Also no need to call `_clear_client()` during `__init__()`.
``` Suggestion
        super().__init__(self.name, config_root=config_root)
        self._load_client()
```

jmartin-tech commented on garak/generators/llm.py:
```diff
@@ -0,0 +1,120 @@
+    Calls model.prompt() with the prompt text and relays the response. Per-provider
+    options and API keys are all handled by `llm` (e.g., `llm keys set openai`).
+
+    Set --model_name to the `llm` model id or alias (e.g., "gpt-4o-mini",
```
```suggestion
    Set --target_name to the `llm` model id or alias (e.g., "gpt-4o-mini",
```

jmartin-tech commented on garak/generators/llm.py:
```diff
@@ -0,0 +1,120 @@
+        if len(system_turns) > 1:
+            raise ValueError("llm generator supports at most one system turn")
+        if len(user_turns) != 1:
+            raise ValueError("llm generator requires exactly one user turn")
```
Unsupported prompts in `_call_model()` are currently expected to return `None` to avoid early termination of the test run. While I understand the thought process of using `raise` here, currently logging the reason for skipping the `prompt` would align better.

```
import logging
```

```suggestion
        if assistant_turns:
            logging.debug("llm generator does not accept assistant turns")
            return [None] * generations_this_call
        if len(system_turns) > 1:
            logging.debug("llm generator supports at most one system turn")
            return [None] * generations_this_call
        if len(user_turns) != 1:
            logging.debug("llm generator requires exactly one user turn")
            return [None] * generations_this_call
```

jmartin-tech commented on garak/generators/llm.py:
```diff
@@ -0,0 +1,120 @@
+            if getattr(self, key) is not None
+        }
+        if self.stop:
+            prompt_kwargs["stop"] = self.stop
```
Could this inspect the accepted arguments to `self.target.prompt()` vs a hard coded list here?  Something similar exists in the OpenAICompatible class, where we collect all options set on the generator that the target API accepts.

jmartin-tech reviewed (changes requested):
Some tweaks to ensure consistent behaviour.

Code suggestions are untested.

jmartin-tech commented on tests/generators/test_llm.py:
```diff
@@ -0,0 +1,173 @@
+# SPDX-FileCopyrightText: Portions Copyright (c) 2025 NVIDIA CORPORATION &
+#                         AFFILIATES. All rights reserved.
+# SPDX-FileCopyrightText: Portions Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Again duplication to be removed.


```suggestion
# SPDX-FileCopyrightText: Portions Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```

Nakul-Rajpal commented:
@jmartin-tech @leondz Hi. I removed the duplicated documentation and license header in the test file. I also fixed the CI test failure. Everything should work now.

Nakul-Rajpal commented:
@leondz @jmartin-tech Just wanted to check whether this can be merged.

jmartin-tech commented on garak/generators/llm.py:
```diff
@@ -0,0 +1,154 @@
+            logging.error(
+                "Failed to resolve `llm` target '%s': %s", self.name, repr(exc)
+            )
+            raise
```
A question, this tool's documentation expects [system or environment stored api keys](https://llm.datasette.io/en/stable/setup.html#api-key-management) however it also looks like it supports passing the api key when a request is made with the `--key` option when accessed as a cli tool.

Given this, should this generator support an `ENV_VAR = "LLM_API_KEY"` that is _optional_ to support configuration of `garak` usage specific keys not previously stored by the system?  If so then, some special handling needs to be added to inject `self.api_key` which is the standard `garak` generator attribute when set as the correct param either utilized in `_load_client()` or `_call_model()` depending on which object from the `llm` package accepts the what the package calls `key` on the cli layer.

leondz replied:
Let's defer to having `llm` manage all its own stuff. Prefer to maintain a minimal interface with third-party tools, reducing our coupling

leondz commented on garak/generators/llm.py:
```diff
@@ -0,0 +1,154 @@
+            return [Message(out)]
+        except Exception as e:
+            logging.error("`llm` generation failed: %s", repr(e))
+            return [None]
```
ignores `generations_this_call`, which should be respected

Nakul-Rajpal commented on garak/probes/encoding.py:
You're right; I was approving changes that weren't associated with this issue. I will be more careful with this in the future. 

I realized I was working off a bad branch, so I created a new branch off main, and moved the code changes only associated with this branch:

garak/generators/llm.py
tests/generators/test_llm.py 
docs/source/garak.generators.llm.rst 

In the future I will read the Garak Docs, and make sure I am not approving AI changes without review.
